File: c++/1422_MaximumScoreAfterSplittingAString.cpp
// Problem 1422: Maximum Score After Splitting a String
// https://leetcode.com/problems/maximum-score-after-splitting-a-string/

#include <string>
#include <algorithm>
using namespace std;

class Solution {
public:
    int maxScore(string s) {
        // 先跑一個迴圈計算所有的1
        int totalOne = 0;
        for (char c : s) {
            if (c == '1') {
                totalOne++;
            }
        }

        int ans = 0;
        int currZero = 0, currOne = 0;

        // 第二個迴圈跑到倒數第一個, 因為切割不能在最後
        for (int i = 0; i < s.size() - 1; ++i) {
            char c = s[i];

            // 統計目前左方所有的 0 和 1 個數
            if (c == '0') {
                currZero++;
            } else {
                currOne++;
            }

            // 在 過去最大結果 和 目前左邊0個數 + 目前右邊1格數 中取最大值
            ans = max(ans, currZero + totalOne - currOne);
        }

        return ans;
    }
};

File: python/1422_maximum_score_after_splitting_a_string.py
# Problem 1422: Maximum Score After Splitting a String
# https://leetcode.com/problems/maximum-score-after-splitting-a-string/

class Solution:
    def max_score(self, s: str) -> int:
        # 先跑一個迴圈計算所有的 1
        total_one = 0
        for c in s:
            if c == '1':
                total_one += 1

        ans = 0
        curr_zero, curr_one = 0, 0
        # 第二個迴圈跑到倒數第一個, 因為切割不能在最後
        for c in s[:-1]:
            # 統計目前左方所有的 0 和 1 個數
            if c == '0':
                curr_zero += 1
            else:
                curr_one += 1
            # 在 過去最大結果 和 目前左邊0個數 + 目前右邊1格數 中取最大值
            ans = max(ans, curr_zero + total_one - curr_one)
        return ans
